### DISCUSSION TITLE
assembling 3rd order tensor ?

### DISCUSSION THREAD

**Discussion by casanovamaxime** [firedrakeproject/firedrake#3529] [Firedrake support] [answered]
Hello i had a little question, is it possible to assemble 3rd order tensor such that the derivative  of my Jacobian J ?
e.g,
J = derivative(F,w) (F my residual and w solution)
JJ = derivative(J,w)
assemble(JJ) gives me errors but assemble(J) works well...

thanks a lot for your nice work and help,

Best,

**Comment by dham** [comment, marked as answer]
No, we (and every other package I am aware of) don't have support for assembling third order tensors. Basically this is because those tensors are huge and hence very expensive to construct and store. Usually your algorithm will only need the action of an order 3 tensor (which is a matrix), and that will work just fine.

**Reply by casanovamaxime** [reply]
thanks for your answer, it is indeed what i have done so far,

however, for another case, i get some shape error when i use the action:
eg:
w = Function(W) (W solution space)
d = Function(D)  (D controle space != W)
J = derivative(F,w) (Jacobian)
JJ = derivative(F,d) (i need to compute the derivative of J with respect to my control d and not w this time)
wadj = Function(W) (adj solution)
act = action(JJ,wadj) gives me shape error which i understand since D!=W but how to get the action in this case ? is it a trick to make match the dimension ?

thanks again for your quick reply,

**Reply by dham** [reply]
OK, so that action doesn't make mathematical sense. `JJ` has three arguments, the test function (in W), trial function (in W) and the direction of the second derivative (in d). Taking the action substitutes into the last argument, which is the one in d. Given that `wadj` is an adjoint variable, I suspect that you actually want to substitute it into the first argument. That would be:
```
derivative(action(adjoint(J), wadj), d) 
```
More generally, I think you want to take the action on wadj before taking the derivative with respect to d (because they are different variables, interchanging this order of operations should be valid).

**Reply by casanovamaxime** [reply]
i see the point, thanks a lot !

**Reply by dham** [reply]
By the way, you may well be making your life unnecessarily hard by writing the adjoint calculation yourself. You know Firedrake has built-in automatic adjoint capabilities, right?

**Reply by casanovamaxime** [reply]
yes indeed i use it !
